File: src/main/java/info/skyblond/archivedag/security/UserDetailService.java
package info.skyblond.archivedag.security;

import info.skyblond.archivedag.model.entity.UserEntity;
import info.skyblond.archivedag.repo.UserRepository;
import info.skyblond.archivedag.repo.UserRoleRepository;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class UserDetailService implements UserDetailsService {
    private final UserRepository userRepository;
    private final UserRoleRepository userRoleRepository;

    public UserDetailService(UserRepository userRepository, UserRoleRepository userRoleRepository) {
        this.userRepository = userRepository;
        this.userRoleRepository = userRoleRepository;
    }

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        UserEntity user = this.userRepository.findByUsername(username);
        if (user == null) {
            throw new UsernameNotFoundException("User not found!");
        }

        List<GrantedAuthority> grantedAuthorities = new ArrayList<>();

        this.userRoleRepository.findAllByUsername(username)
                .stream().map(role -> new SimpleGrantedAuthority(role.getRole()))
                .forEach(grantedAuthorities::add);

        return new User(user.getUsername(), user.getPassword(),
                user.getStatus() == UserEntity.UserStatus.ENABLED,
                true, true,
                user.getStatus() != UserEntity.UserStatus.LOCKED,
                grantedAuthorities);
    }
}

File: backend/src/main/kotlin/info/skyblond/archivedag/arudaz/security/UserDetailService.kt
package info.skyblond.archivedag.arudaz.security

import info.skyblond.archivedag.arstue.entity.UserEntity
import info.skyblond.archivedag.arstue.repo.UserRepository
import info.skyblond.archivedag.arstue.repo.UserRoleRepository
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

@Service
class UserDetailService(
    private val userRepository: UserRepository,
    private val userRoleRepository: UserRoleRepository
) : UserDetailsService {

    override fun loadUserByUsername(username: String): UserDetails {
        val user = userRepository.findByUsername(username) ?: throw UsernameNotFoundException("User not found!")
        val grantedAuthorities: MutableList<GrantedAuthority> = ArrayList()
        userRoleRepository.findAllByUsername(username)
            .map { SimpleGrantedAuthority(it.role) }
            .forEach { grantedAuthorities.add(it) }
        return User(
            user.username, user.password,
            user.status === UserEntity.Status.ENABLED,
            true, true,
            user.status !== UserEntity.Status.LOCKED,
            grantedAuthorities
        )
    }
}
